### ISSUE TITLE
For external Bazel projects, java_libraries are attached with mismatched bytes/src

### ISSUE THREAD

**Issue by dhalperi** [bazelbuild/intellij#1640]
# tl;dr
Repro in this repo: https://github.com/dhalperi/bazel-intellij-hjar

Import the project (default view), open `DemoClass`, navigate to `BDD` sources. They load, but then there's a banner that there's a binary mismatch [1]. It appears to be caused by comparing the real source to the bdd hjar.

I don't know if this is a bug or WAI.

# Metadata

```
Bazelisk version: v1.3.0
Build label: 2.1.0
Build target: bazel-out/darwin-opt/bin/src/main/java/com/google/devtools/build/lib/bazel/BazelServer_deploy.jar
Build time: Fri Feb 7 13:04:00 2020 (1581080640)
Build timestamp: 1581080640
Build timestamp as int: 1581080640
```

Using IntelliJ 2019.2.4 and Bazel plugin 2020.01.28.0.3

# Overview

The referenced project has a `git_repository` dependence on  another Bazel project, plus `rules_jvm_external` jars from Maven.

If you import the project (default view is fine), you will see that it builds fine. If you jump to the definition of `ImmutableList` or `StringUtils`, you'll see the right source, javadoc, no issues.

If you jump to the definition of `BDD`, which is a `java_library` in the imported `batfish2` Bazel project, you will see the right source, plus this banner: `Library source does not match the bytecode for class BDD` [1]. Looking at the diff [2] makes it obvious that it's comparing the `hjar` to the real sources. We also see this from the external libraries list [3,4].

Note that the `Use a local jar cache` setting does not appear to affect this issue at all.

Is this a bug or WAI, and we have to deal with it?

# Screenshots

[1] <img width="1215" alt="Screen Shot 2020-02-13 at 9 24 28 PM" src="https://user-images.githubusercontent.com/526415/74503879-41e12300-4ea7-11ea-8890-44636bcb193d.png">

[2] <img width="1680" alt="Screen Shot 2020-02-13 at 9 23 18 PM" src="https://user-images.githubusercontent.com/526415/74503831-12cab180-4ea7-11ea-93e5-32b8a58f4703.png">

[3] <img width="330" alt="Screen Shot 2020-02-13 at 9 25 31 PM" src="https://user-images.githubusercontent.com/526415/74503942-6d640d80-4ea7-11ea-8598-a78a23158991.png">

[4] <img width="1012" alt="Screen Shot 2020-02-13 at 9 25 40 PM" src="https://user-images.githubusercontent.com/526415/74503947-7359ee80-4ea7-11ea-83db-f1fb7836bf77.png">

**Comment by dhalperi** [comment]
# IDE Bazel sync log, after a few restarts:

```
Syncing project: Sync (incremental)...
Updating VCS...
Running Bazel info...
Command: /usr/local/bin/bazelisk info --tool_tag=ijwb:IDEA:community --curses=no --color=yes --progress_in_terminal_title=no --

INFO: Invocation ID: 1f7819a9-1603-493e-85b1-c506932a2c4d
Computing VCS working set...
Querying targets in project directories...
Command: /usr/local/bin/bazelisk query --tool_tag=ijwb:IDEA:community --output=label_kind --keep_going "attr(\"tags\", \"^((?!manual).)*$\", //...:all - //bazel-bin/...:all - //bazel-genfiles/...:all - //bazel-out/...:all - //bazel-testlogs/...:all - //bazel-intellij-demo/...:all - //.ijwb/...:all)" --

1 targets found under project directories; syncing 1 of them.
Sync targets from project view directories:
  //:demo

Building Bazel targets...
Command: /usr/local/bin/bazelisk build --tool_tag=ijwb:IDEA:community --keep_going --build_event_binary_file=/var/folders/qb/g94r9sq10ms05d6gskv1g1l80000gn/T/intellij-bep-c42d2729-fa46-4dc8-89d0-808bed7671fe --nobuild_event_binary_file_path_conversion --curses=no --color=yes --progress_in_terminal_title=no --aspects=@intellij_aspect//:intellij_info_bundled.bzl%intellij_info_aspect "--override_repository=intellij_aspect=/Users/dan/Library/Application Support/IdeaIC2019.2/ijwb/aspect" --output_groups=intellij-info-generic,intellij-info-java-direct-deps,intellij-resolve-java-direct-deps -- //:demo

INFO: Invocation ID: 79f67fb4-e50d-467a-b2dd-4f8bca5d7d98
Loading: 
Loading: 0 packages loaded
Analyzing: target //:demo (0 packages loaded, 0 targets configured)
INFO: Analyzed target //:demo (1 packages loaded, 12 targets configured).
INFO: Found 1 target...
[0 / 31] [Prepa] BazelWorkspaceStatusAction stable-status.txt
Aspect @intellij_aspect//:intellij_info_bundled.bzl%intellij_info_aspect of //:demo up-to-date:
  bazel-bin/external/bazel_tools/tools/jdk/current_java_toolchain--1286701614.intellij-info.txt
  bazel-bin/external/batfish2/projects/bdd/bdd.java-manifest
  bazel-bin/external/batfish2/projects/bdd/bdd-97378.intellij-info.txt
  bazel-bin/external/maven/com_google_guava_guava-46378293.intellij-info.txt
  bazel-bin/external/maven/org_apache_commons_commons_lang3--1479187423.intellij-info.txt
  bazel-bin/demo.java-manifest
  bazel-bin/demo-3079651.intellij-info.txt
  bazel-bin/external/batfish2/projects/bdd/libbdd.jar
  bazel-bin/external/batfish2/projects/bdd/libbdd-hjar.jar
  bazel-bin/external/batfish2/projects/bdd/libbdd-src.jar
  bazel-bin/external/batfish2/projects/bdd/libbdd.jdeps
  bazel-bin/external/maven/v1/https/repo1.maven.org/maven2/com/google/guava/guava/28.1-jre/stamped_guava-28.1-jre.jar
  bazel-bin/external/maven/v1/https/repo1.maven.org/maven2/com/google/guava/guava/28.1-jre/guava-28.1-jre-sources.jar
  bazel-bin/external/maven/v1/https/repo1.maven.org/maven2/org/apache/commons/commons-lang3/3.9/stamped_commons-lang3-3.9.jar
  bazel-bin/external/maven/v1/https/repo1.maven.org/maven2/org/apache/commons/commons-lang3/3.9/commons-lang3-3.9-sources.jar
  bazel-bin/libdemo.jar
  bazel-bin/libdemo-hjar.jar
  bazel-bin/libdemo-src.jar
  bazel-bin/libdemo.jdeps
  bazel-bin/external/maven/v1/https/repo1.maven.org/maven2/com/google/code/findbugs/jsr305/3.0.2/stamped_jsr305-3.0.2.jar
  bazel-bin/external/maven/v1/https/repo1.maven.org/maven2/com/google/guava/listenablefuture/9999.0-empty-to-avoid-conflict-with-guava/stamped_listenablefuture-9999.0-empty-to-avoid-conflict-with-guava.jar
  bazel-bin/external/maven/v1/https/repo1.maven.org/maven2/org/codehaus/mojo/animal-sniffer-annotations/1.18/stamped_animal-sniffer-annotations-1.18.jar
  bazel-bin/external/maven/v1/https/repo1.maven.org/maven2/com/google/j2objc/j2objc-annotations/1.3/stamped_j2objc-annotations-1.3.jar
  bazel-bin/external/maven/v1/https/repo1.maven.org/maven2/org/checkerframework/checker-qual/2.8.1/stamped_checker-qual-2.8.1.jar
  bazel-bin/external/maven/v1/https/repo1.maven.org/maven2/com/google/guava/failureaccess/1.0.1/stamped_failureaccess-1.0.1.jar
  bazel-bin/external/maven/v1/https/repo1.maven.org/maven2/com/google/errorprone/error_prone_annotations/2.3.2/stamped_error_prone_annotations-2.3.2.jar
  bazel-bin/external/maven/v1/https/repo1.maven.org/maven2/com/google/code/findbugs/jsr305/3.0.2/jsr305-3.0.2-sources.jar
  bazel-bin/external/maven/v1/https/repo1.maven.org/maven2/org/codehaus/mojo/animal-sniffer-annotations/1.18/animal-sniffer-annotations-1.18-sources.jar
  bazel-bin/external/maven/v1/https/repo1.maven.org/maven2/com/google/j2objc/j2objc-annotations/1.3/j2objc-annotations-1.3-sources.jar
  bazel-bin/external/maven/v1/https/repo1.maven.org/maven2/org/checkerframework/checker-qual/2.8.1/checker-qual-2.8.1-sources.jar
  bazel-bin/external/maven/v1/https/repo1.maven.org/maven2/com/google/guava/failureaccess/1.0.1/failureaccess-1.0.1-sources.jar
  bazel-bin/external/maven/v1/https/repo1.maven.org/maven2/com/google/errorprone/error_prone_annotations/2.3.2/error_prone_annotations-2.3.2-sources.jar
INFO: Elapsed time: 1.926s, Critical Path: 1.32s
INFO: 2 processes: 2 darwin-sandbox.
INFO: Build completed successfully, 3 total actions
INFO: Build completed successfully, 3 total actions
INFO: Build Event Protocol files produced successfully.
INFO: Build completed successfully, 3 total actions
build invocation result: success
Parsing build outputs...
Total rules: 5, new/changed: 0, removed: 0
Reading IDE info result...
Updating target map
Loaded 0 aspect files, total size 0kB
Target map size: 5
Reading jdeps files...
Loaded 0 jdeps files, total size 0kB
Reading package manifests...
Java content entry count: 1
Updating Jar Cache...
Prefetching files...
Refreshing files
Computing directory structure...
Initializing project SDKs...
Committing project structure...
Workspace has 3 libraries
Workspace has 2 modules
Updating in-memory state...
Sync finished

Timing summary:
BlazeInvocation: 2.6s, Prefetching: 0ms, Other: 105ms
```

**Comment by dhalperi** [comment]
# the intellij-info.txt for bdd

```
build_file_artifact_location {
  is_external: true
  is_new_external_version: true
  is_source: true
  relative_path: "projects/bdd/BUILD"
  root_execution_path_fragment: "external/batfish2"
}
deps {
  dependency_type: 0
  target {
    label: "@bazel_tools//tools/jdk:current_java_toolchain"
  }
}
deps {
  dependency_type: 0
  target {
    label: "@maven//:com_google_code_findbugs_jsr305"
  }
}
java_ide_info {
  jars {
    interface_jar {
      is_external: true
      is_new_external_version: true
      is_source: false
      relative_path: "external/batfish2/projects/bdd/libbdd-hjar.jar"
      root_execution_path_fragment: "bazel-out/darwin-fastbuild/bin"
    }
    jar {
      is_external: true
      is_new_external_version: true
      is_source: false
      relative_path: "external/batfish2/projects/bdd/libbdd.jar"
      root_execution_path_fragment: "bazel-out/darwin-fastbuild/bin"
    }
    source_jar {
      is_external: true
      is_new_external_version: true
      is_source: false
      relative_path: "external/batfish2/projects/bdd/libbdd-src.jar"
      root_execution_path_fragment: "bazel-out/darwin-fastbuild/bin"
    }
    source_jars {
      is_external: true
      is_new_external_version: true
      is_source: false
      relative_path: "external/batfish2/projects/bdd/libbdd-src.jar"
      root_execution_path_fragment: "bazel-out/darwin-fastbuild/bin"
    }
  }
  jdeps {
    is_external: true
    is_new_external_version: true
    is_source: false
    relative_path: "external/batfish2/projects/bdd/libbdd.jdeps"
    root_execution_path_fragment: "bazel-out/darwin-fastbuild/bin"
  }
  package_manifest {
    is_external: true
    is_new_external_version: true
    is_source: false
    relative_path: "external/batfish2/projects/bdd/bdd.java-manifest"
    root_execution_path_fragment: "bazel-out/darwin-fastbuild/bin"
  }
  sources {
    is_external: true
    is_new_external_version: true
    is_source: true
    relative_path: "projects/bdd/src/main/java/net/sf/javabdd/BDD.java"
    root_execution_path_fragment: "external/batfish2"
  }
  sources {
    is_external: true
    is_new_external_version: true
    is_source: true
    relative_path: "projects/bdd/src/main/java/net/sf/javabdd/BDDBitVector.java"
    root_execution_path_fragment: "external/batfish2"
  }
  sources {
    is_external: true
    is_new_external_version: true
    is_source: true
    relative_path: "projects/bdd/src/main/java/net/sf/javabdd/BDDDomain.java"
    root_execution_path_fragment: "external/batfish2"
  }
  sources {
    is_external: true
    is_new_external_version: true
    is_source: true
    relative_path: "projects/bdd/src/main/java/net/sf/javabdd/BDDException.java"
    root_execution_path_fragment: "external/batfish2"
  }
  sources {
    is_external: true
    is_new_external_version: true
    is_source: true
    relative_path: "projects/bdd/src/main/java/net/sf/javabdd/BDDFactory.java"
    root_execution_path_fragment: "external/batfish2"
  }
  sources {
    is_external: true
    is_new_external_version: true
    is_source: true
    relative_path: "projects/bdd/src/main/java/net/sf/javabdd/BDDPairing.java"
    root_execution_path_fragment: "external/batfish2"
  }
  sources {
    is_external: true
    is_new_external_version: true
    is_source: true
    relative_path: "projects/bdd/src/main/java/net/sf/javabdd/JFactory.java"
    root_execution_path_fragment: "external/batfish2"
  }
}
key {
  label: "@batfish2//projects/bdd:bdd"
}
kind_string: "java_library"
```

**Comment by dhalperi** [comment]
FWIW, the debugger seems to work but the banner is present the entire time on most classes (the ones where they have private members or functions).

<img width="1680" alt="Screen Shot 2020-02-13 at 9 42 19 PM" src="https://user-images.githubusercontent.com/526415/74504648-bae17a00-4ea9-11ea-9376-b55967ecb413.png">

(I did have to change the build slightly to be able to run the class, of course).

**Comment by brendandouglas** [comment]
This is basically WAI. We index header jars for better performance, but by design they differ from the source jars according to IntelliJ's criteria (LibrarySourceNotificationProvider).

If you find a way to disable that editor notification, we'd be happy to accept a PR for it.

**Comment by dhalperi** [comment]
rad, thanks!
